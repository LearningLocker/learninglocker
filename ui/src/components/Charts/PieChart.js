import React from 'react';
import { PieChart as Chart, Pie, Tooltip, Cell } from 'recharts';
import { AutoSizer } from 'react-virtualized';
import NoData from 'ui/components/Graphs/NoData';
import getColor from 'ui/utils/colors';
import {
  getResultsData,
  getChartData,
  hasData,
  getLongModel,
} from './Chart';
import { Chart as StyledChart, } from './styled';
import PieTooltip from './PieTooltip';

const cellPadding = 5;

const getInnerRadius = size => i => hole => (
  (size * i) + (cellPadding * i) + hole
);

const getOuterRadius = size => i => hole =>
  (size * (i + 1)) + (cellPadding * i) + hole;

const getPieSize = (cells, maxSize = 375, hole = 0) => {
  const widgetSize = maxSize;
  return ((widgetSize / 2.7) - hole) / cells;
};

const sortData = data =>
  data.sortBy(e => e.get('total'));

const getSortedData = results => labels =>
  sortData(getResultsData(results)(labels));

const renderCell = numberOfEntries => color => (entry, i) => (
  <Cell
    fill={getColor({
      hexColor: color,
      range: numberOfEntries,
      value: i
    })}
    key={`cell-${i}`} />
  );

const renderTooltips = labels => data => count => grouping => (<Tooltip content={<PieTooltip display={getLongModel(data)} labels={labels} count={count} grouping={grouping} />} />);

const renderPie = colors => data => maxSize => isDonut => (label, i, labels) => {
  const holeWidth = 60;
  const hole = isDonut ? holeWidth : 0;

  return (
    <Pie
      key={i}
      data={getChartData(data)}
      nameKey={'_id'}
      dataKey={`Series ${i + 1}`}
      innerRadius={getInnerRadius(getPieSize(labels.size, maxSize, hole))(i)(hole)}
      outerRadius={getOuterRadius(getPieSize(labels.size, maxSize, hole))(i)(hole)}>
      {data.valueSeq().map(renderCell(data.size)(colors.get(i)))}
    </Pie>
  );
};
const renderPies = labels => colors => data => isDonut => maxSize => labels.map(renderPie(colors)(data)(maxSize)(isDonut)).valueSeq();

const renderPieChart = labels => colors => data => count => grouping => isDonut => ({ width, height }) => (
  <Chart width={width} height={height}>
    {renderPies(labels)(colors)(data)(isDonut)(Math.min(width, height))}
    {renderTooltips(labels)(data)(count)(grouping)}
  </Chart>
);

const renderChartResults = labels => results => colors => count => grouping => isDonut =>
  renderPieChart(labels)(colors)(getSortedData(results)(labels))(count)(grouping)(isDonut);

const renderResults = results => labels => colors => count => grouping => isDonut => (
  <StyledChart>
    <div>
      {(colors.size > 1) && colors.map((colour, i) => (
        <div key={i} style={{ fontSize: '0.9em' }}>
          <div
            style={{ display: 'inline-block', width: '12px', height: '12px', backgroundColor: colour, marginRight: '8px', marginBottom: '2px', verticalAlign: 'middle' }} />
          <span title={labels.get(i) || `Series ${i + 1}`}>{(labels.get(i) && labels.get(i).substr(0, 45)) || `Series ${i + 1}`}</span>
        </div>)
      )}
    </div>
    <div style={{ height: '100%', width: '100%' }}>
      <AutoSizer>{renderChartResults(labels)(results)(colors)(count)(grouping)(isDonut)}</AutoSizer>
    </div>
  </StyledChart>
);

export const PieChartComponent = ({ results, labels, colors, count, grouping, isDonut }) =>
(hasData(results) ? renderResults(results)(labels)(colors)(count)(grouping)(isDonut) : <NoData />);

export default PieChartComponent;
