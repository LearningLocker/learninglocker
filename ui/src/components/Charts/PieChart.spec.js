import React from 'react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import { List, fromJS } from 'immutable';
import PieChart from './PieChart';


describe('PieChart', () => {
  test('should render no data', () => {
    const pieChart = renderer
      .create(<PieChart results={new List()} />)
      .toJSON();

    expect(pieChart).toMatchSnapshot();
  });

  test('should render', () => {
    const results = fromJS([[{
      0: {
        _id: 0,
        count: 1,
        model: 0
      },
      1: {
        _id: 1,
        count: 5,
        model: 1
      }
    }]]);

    const pieChart = renderer
      .create(<PieChart
        results={results}
        labels={fromJS([undefined])}
        colors={fromJS(['#ffffff', '#123456'])} />)
      .toJSON();

    expect(pieChart).toMatchSnapshot();
  });
});
