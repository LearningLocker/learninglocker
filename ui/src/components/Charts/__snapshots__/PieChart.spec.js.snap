// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart should render 1`] = `
.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 svg g text {
  font-size: 12px;
}

<div
  className="c0"
>
  <div>
    <div
      style={
        Object {
          "fontSize": "0.9em",
        }
      }
    >
      <div
        style={
          Object {
            "backgroundColor": "#ffffff",
            "display": "inline-block",
            "height": "12px",
            "marginBottom": "2px",
            "marginRight": "8px",
            "verticalAlign": "middle",
            "width": "12px",
          }
        }
      />
      <span
        title="Series 1"
      >
        Series 1
      </span>
    </div>
    <div
      style={
        Object {
          "fontSize": "0.9em",
        }
      }
    >
      <div
        style={
          Object {
            "backgroundColor": "#123456",
            "display": "inline-block",
            "height": "12px",
            "marginBottom": "2px",
            "marginRight": "8px",
            "verticalAlign": "middle",
            "width": "12px",
          }
        }
      />
      <span
        title="Series 2"
      >
        Series 2
      </span>
    </div>
  </div>
  <div
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <div
      className={undefined}
      style={
        Object {
          "height": 0,
          "overflow": "visible",
          "width": 0,
        }
      }
    />
  </div>
</div>
`;

exports[`PieChart should render no data 1`] = `
<h4
  style={
    Object {
      "marginTop": 20,
      "textAlign": "center",
      "width": "100%",
    }
  }
>
  No data in this time range.
</h4>
`;
